presentation/backend/src/presentation-backend/ElementPropertiesHelper.ts

```diff
@@ -37,70 +34,285 @@ function getECSqlName(fullClassName: string) {
+                selectClassName: classFullName,
+                expression: createElementIdsECExpressionFilter(batch),
+              };
+              return contentSetGetter({
+                rulesetOrId: ruleset,
+                keys,
+                descriptor: filteringDescriptor,
+              });
+            }).pipe(map((items) => ({ descriptor, items }))),
+          batchesParallelism,
+        ),
+      ),
+    ),
+  );
+}
+
+function createElementIdsECExpressionFilter(batch: Array<{ from: Id64String; to: Id64String }>): string {
+  let filter = "";
+  function appendCondition(cond: string) {
+    if (filter.length > 0) {
+      filter += " OR ";
     }
+    filter += cond;
   }
-  if (fromId && currId) {
-    batches.push({ from: fromId, to: currId });
+  for (const item of batch) {
+    if (item.from === item.to) {
+      appendCondition(`this.ECInstanceId = ${item.from}`);
+    } else {
+      appendCondition(`this.ECInstanceId >= ${item.from} AND this.ECInstanceId <= ${item.to}`);
+    }
   }
-  return batches;
+  return filter;
 }
 
-function createElementsFilter(elementAlias: string, classNames?: string[]) {
-  if (classNames === undefined || classNames.length === 0) {
-    return undefined;
-  }
+function createClassContentRuleset(fullClassName: string): Ruleset {
+  const [schemaName, className] = parseFullClassName(fullClassName);
+  return {
+    id: `content/class-descriptor/${fullClassName}`,
+    rules: [
+      {
+        ruleType: "Content",
+        specifications: [
+          {
+            specType: "ContentInstancesOfSpecificClasses",
+            classes: {
+              schemaName,
+              classNames: [className],
+              arePolymorphic: false,
+            },
+            handlePropertiesPolymorphically: true,
+          },
+        ],
+      },
+    ],
+  };
+}
 
-  // check if list contains only valid class names
-  const classNameRegExp = new RegExp(/^[\w]+[.:][\w]+$/);
-  const invalidName = classNames.find((name) => !name.match(classNameRegExp));
-  if (invalidName) {
-    throw new PresentationError(
-      PresentationStatus.InvalidArgument,
-      `Encountered invalid class name - ${invalidName}.
-      Valid class name formats: "<schema name or alias>.<class name>", "<schema name or alias>:<class name>"`,
-    );
-  }
+/** Given a list of element ids, group them by class name. */
+function getElementClassesFromIds(imodel: IModelDb, elementIds: string[]): Observable<{ classFullName: string; ids: Id64Array }> {
+  const elementIdsBatchSize = 5000;
+  return range(0, elementIds.length / elementIdsBatchSize).pipe(
+    mergeMap((batchIndex) => {
+      const idsFrom = batchIndex * elementIdsBatchSize;
+      const idsTo = Math.min(idsFrom + elementIdsBatchSize, elementIds.length);
+      let idsFilter = "";
+      for (let i = idsFrom; i < idsTo; i++) {
+        idsFilter += `${elementIds[i]}`;
+        if (i < idsTo - 1) {
+          idsFilter += ",";
+        }
+      }
```

Comment:
Is this faster than `slice` and `join`? I thought that concatenating string in js is not a cheap operation.

Reply:
You're right - that's ~2x faster. Changed.